secure_delete/srm.1
.\" This definition swiped from the gcc(1) man page
.de Sp
.if n .sp
.if t .sp 0.4
..
.TH SRM 1

.SH NAME
srm \- secure remove (secure_deletion toolkit)

.SH SYNOPSIS
.B srm [-d] [-f] [-l] [-l] [-r] [-v] [-z] files

.SH DESCRIPTION
.I srm 
is designed to delete data on mediums in a secure manner which can not be
recovered by thiefs, law enforcement or other threats.
The wipe algorythm is based on the paper "Secure Deletion of Data from
Magnetic and Solid-State Memory" presented at the 6th Usenix Security
Symposium by Peter Gutmann, one of the leading civilian cryptographers.
.PP
The
.I secure data deletion
process of srm goes like this:
.PP
.TP
.B *
1 pass with 0xff
.TP
.B * 
5 random passes. /dev/urandom is used for a secure RNG if available.
.TP
.B * 
27 passes with special values defined by Peter Gutmann.
.TP
.B * 
5 random passes. /dev/urandom is used for a secure RNG if available.
.TP
.B *   
Rename the file to a random value
.TP
.B * 
Truncate the file
.PP
.PP
As an additional measure of security, the file is opened in O_SYNC mode
and after each pass an fsync() call is done.
.I srm
writes 32k blocks for the purpose of speed, filling buffers of disk caches
to force them to flush and overwriting old data which belonged to the file.
.PP

.SH COMMANDLINE OPTIONS
.PP
.TP
.B \-d
ignore the two special dot files . and .. on the commandline. (so you can
execute it like "srm -d .* *")
.TP
.B \-f
fast (and insecure mode): no /dev/urandom, no synchronize mode.
.TP
.B \-l 
lessens the security. Only two passes are written: one mode with 0xff
and a final mode random values.
.TP
.B \-l
-l for a second time lessons the security even more: only one random pass
is written.
.TP
.B \-r
recursive mode, deletes all subdirectories.
.TP
.B \-v
verbose mode
.TP
.B \-z
wipes the last write with zeros instead of random data
.PP

.SH LIMITATIONS
.TP
.B NFS
Beware of NFS. You can't ensure you really completely wiped your data
from the remote disks.
.TP
.B Raid
Raid Systems use stripped disks and have got large caches. It's hard to wipe
them.
.TP
.B swap, /tmp, etc.
Some of your data might have a temporary (deleted) copy somewhere on the
disk. You should use
.I sfill
which comes with the
.I secure_deletion package
to ensure to wipe also the free diskspace. However, If already a small
file aquired a block with your precious data, no tool known to me can help
you here. For a secure deletion of the swap space
.I sswap
is available.

.PP
.SH BUGS 
No bugs. There was never a bug in the secure_deletion package (in contrast
to my other tools, whew, good luck ;-)
Send me any that you find.  Patches are nice too :)

.SH AUTHOR
.Sp
van Hauser / THC
.I <vh@thc.org>

.SH DISTRIBUTION
The newest version of the
.I secure_deletion package
can be obtained from 
.I http://www.thc.org
.Sp
.I srm
and the
.I secure_deletion package
is (C) 1997-2003 by van Hauser / THC (vh@thc.org)
.Sp
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; Version 2.
.Sp
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

.SH SEE ALSO
.I sfill
(1),
.I sswap
(1),
.I smem
(1)


secure_delete/CHANGES
v3.1 (November 2003)	PUBLIC RELEASE
	(sdel-mod)	I added a kernel module in v3.0 which wipes all
			files which are unlinked() in the kernel. However
			it didnt work, so it was not official.
			However, deathcubek@md2600.net sent in a patch,
			so now it is working :-)
			After "make install" do "insmod sdel-mod"
	(srm)		Fixed a bug which prevented to delete directories

v3.0 (April 2003)	PUBLIC RELEASE
	(all)		Added >2GB file support
	(all)		rewrote all tools to use a central library, this
			was long time necessary!
	(all)		warning now displayed if ulimit value resetting
			support was not compiled in.
	(sswap)		added -j(ump) option to skip over a defined number
			of bytes of the block device.
	(srm)		skipping file wipe if it is hardlinked!
	(fileutils-rm)	Update the rm.diff for the newest fileutil (4.1)
	(all)		updated the link for our homepage + email to thc.org

v2.5 (May 2001)		PUBLIC RELEASE
	(all)		updated the link for our homepage to
			www.thehackerschoice.com

v2.4 (October 2000)	PRIVATE RELEASE
	(all)		added the option -z to write zeros as last wipe
			instead of random data
	(sfill)		added the options -i and -I to not or exclusivly wipe
                        inode space and expanded the number of inodes to wipe
	(the_cleaner.sh) added a new shellscript called the_cleaner.sh which
			wipes temporary directories, all user's history and
			netscape history/cache files, free disk and inode
			space, swap space and the memory. Easy shellscript
			to place in the shutdown phase of runlevel 2 and 3!
	(all)		beautified the help output; updated documentation
	(all)		added srm/sfill/sswap/smem as linux elf binaries

v2.3 (August 2000)	PUBLIC RELEASE
	(sfill)		now also wipes all inodes in the directory specified
	(srm)		fixed a bug in srm when deleting dirs without -r option
			(thanks to typo/teso)

v2.2 (October 1999)	PUBLIC RELEASE
	(WWW)		ugh, the redirector service base.org is down!
			WWW site: http://www.infowar.co.uk/thc
	(all)		code cleanup to prevent warnings (thanks Rebecca!)
			(some weird systems need O_FSYNC and strings.h)
	(srm)		the blocksize used for wiping is now enlarged if the
			filesystem is detected to use bigger blocks.
	(srm/sfill/sswap) changed the behaviour, now the last wipe is always
			a random one, so it's harder to find out that a
			secure deletion was done. (Thanks to Dan Farmer for
			this idea)
	(srm)		the random buffer to write was not renewed after one
			write to the file, this makes it pretty fast but also
			not that secure. No all random writes are unique. The
			other tools had that from the beginning on. duh

v2.1 (April 1999)	PRIVATE RELEASE
	(all)		added the -f(ast) option.
	(smem)		added a new tool: smem, which wipes free memory
	(srm/sfill/smem) added rlimit resets
	(sfill/sswap)	quadroupled the buffer size
	(srm/sfill/sswap) fix for openbsd which doesn't know O_SYNC open(2).
        (srm)		added the -d(ot) option. Ignores "." and ".." on the
			commandline
	(srm)		fixed directory and file link races.
			Thanks to Solar Designer bringing this to my attention.
			Note that there is a small racecondition left which
			can be exploited from an attacker to chmod 0600 a file
			to it's owner if the owner is running srm and the
			attacker has write access to the target directory.
			The other wipe programs have this problem too - and
			( - wiping anything in your tree for which you
			have write permission ... - ) worse.

v2.0
	(all)		added man pages for srm, sfill and sswap!
	(all)		increased speed by ~300% by writing big buffers.
	(all)		using /dev/urandom for random passes now if available
			(this is more secure, however costs speed)
	(all)		added the O_SYNC bit to all open calls. Should not
			add security, but who knows how weired systems behave.
	(all)		the -s option(s) is now ignored. The standard mode
			is now the secure mode. Use -l to lessen the security.
	(srm)		added the option -r for directory recursion
	(srm)		doing a chmod() now before opening the file
	(srm)		fixed handling of symlink and special files
	(rm.diff)	includes all the above except for the new -s/-l change
	(all)		cosmetics + documenation updates

v1.8
	(rm.diff)	updated the diff and did some more security and speed.
	(all)		a bit cosmetics

v1.7
	(srm)		using fsync() now, which gives you 10-25% speed !
	(all)		using perror() for better error reporting
	(srm/sfill)	included file truncating before unlinking (security!)

v1.6
	[www distrib.]	http://r3wt.base.org - check out the THC area
	[email]		my new email: vh@reptile.rug.ac.be
	(sswap)		added a new tool: a secure swap cleaner. remember to
			unmount your swap first. *only tested on linux* !
	(srm/sfill) 	added trapping signals and exiting clean on them
	(srm/sfill)	errors are sent do stderr now instead to stdout
	(srm/sfill)	general code cleanup
	(sfill)		fixed race condition to prevent symlink attacks
	(sfill)		warning if executed by user!=root, because not all
			free space can be overwritten (idea by fyodor@dhp.com)
	(rm.diff)	replaces the rm.c replacement with a diff (+ bugfix)
	(sdel.exe/sfill.exe) removed the msdos versions. They 1st already
			included the features above and 2nd are no longer
			supported (at least by me - get a good unix)

v1.4
	(sfill)		some flushing on stdout done
	(sfill)		more error checking
	(srm/sfill)	cosmetics

v1.3
			initial public release.


secure_delete/srm.c
/* Secure RM - by van Hauser / [THC], vh@thc.org
 *
 * Secure ReMove first overwrites then renames and finally deletes the target
 * file(s) specified via parameters.
 * For security reasons full 32kb blocks are written so that the whole block
 * on which the file(s) live are overwritten. (change #define #BLOCKSIZE)
 * The option -l overwrites two times the data.
 * The option -ll overwrites the data once.
 * Standard mode is a real security wipe for 38 times, flushing
 * the caches after every write. The wipe technique was proposed by Peter
 * Gutmann at Usenix '96 and includes 10 random overwrites plus 28 special
 * defined characters. Take a look at the paper of him, it's really worth
 * your time. 
 *
 * Advice : set "alias rm 'srm -v'"
 *
 * Read the manual for limitations.
 * Compiles clean on OpenBSD, Linux, Solaris, AIX and I guess all others.
 *
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <strings.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <dirent.h>
#include <signal.h>
#include <sys/time.h>
#include <sys/resource.h>

#include "sdel.h"
   
int slow = O_SYNC;
int recursive = 0;
int zero = 0;
unsigned long bufsize = BLOCKSIZE;
int fd;

void help () {
    printf("srm %s (c) 1997-2003 by %s <%s>\n\n", VERSION, AUTHOR, EMAIL);
    printf("Syntax: %s [-dflrvz] file1 file2 etc.\n\n", prg);
    printf("Options:\n");
    printf("\t-d  ignore the two dot special files \".\" and \"..\".\n");
    printf("\t-f  fast (and insecure mode): no /dev/urandom, no synchronize mode.\n");
    printf("\t-l  lessens the security (use twice for total insecure mode).\n");
    printf("\t-r  recursive mode, deletes all subdirectories.\n");
    printf("\t-v  is verbose mode.\n");
    printf("\t-z  last wipe writes zeros instead of random data.\n");
    printf("\nsrm does a secure overwrite/rename/delete of the target file(s).\n");
    printf("Default is secure mode (38 writes).\n");
    printf("You can find updates at %s\n", WEB);
    exit(1);
}

int smash_it(char *filename, int mode)  {
    struct stat filestat;
    struct stat controlstat;
    int i_am_a_directory = 0;

/* get the file stats */
    if (lstat(filename, &filestat))
            return 1;

    if (S_ISREG(filestat.st_mode) && filestat.st_nlink > 1) {
        fprintf(stderr, "Error: File %s - file is hardlinked %d time(s), skipping!\n", filename, filestat.st_nlink - 1);
        return -1;
    }

/* if the blocksize on the filesystem is bigger than the on compiled with, enlarge! */
    if (filestat.st_blksize > bufsize) {
	if (filestat.st_blksize > 65532) {
	    bufsize = 65535;
	} else {
	    bufsize = (((filestat.st_blksize / 3) + 1) * 3);
	}
    }

/* handle the recursive mode */
    if (recursive)
        if (S_ISDIR(filestat.st_mode)) {
	    DIR *dir;
	    struct dirent *dir_entry;
	    struct stat cwd_stat;
	    char current_dir[4097];
	    int res;
	    int chdir_success = 1;
 
            if (verbose) printf("DIRECTORY (going recursive now)\n");
            getcwd(current_dir, 4096);
            current_dir[4096] = '\0';
            
            /* a won race will chmod a file to 0700 if the user is owner/root
               I'll think about a secure solution to this, however, I think
               there isn't one - anyone with an idea? */
            if (chdir(filename)) {
                (void) chmod(filename, 0700); /* ignore permission errors */
                if (chdir(filename)) {
                    fprintf(stderr,"Can't chdir() to %s, hence I can't wipe it.\n", filename);
                    chdir_success = 0;
                }
            }

	    if (chdir_success) {
	        lstat(".", &controlstat);
	        lstat("..", &cwd_stat);
	        if ( (filestat.st_dev != controlstat.st_dev) || (filestat.st_ino != controlstat.st_ino) ) {
	            fprintf(stderr, "Race found! (directory %s became a link)\n", filename);
	        } else {
  	            if ((dir = opendir (".")) != NULL) {
	                (void) chmod(".", 0700); /* ignore permission errors */
	                dir = opendir (".");
	            }
		    if (dir != NULL) {
		        while ((dir_entry = readdir(dir)) != NULL)
			    if (strcmp(dir_entry->d_name, ".") && strcmp(dir_entry->d_name, "..")) {
			        if (verbose) printf("Wiping %s ", dir_entry->d_name);
		  	        if ( (res = smash_it(dir_entry->d_name, mode)) > 0) {
		  	            if (res == 3)
		  	                fprintf(stderr,"File %s was raced, hence I won't wipe it.\n", dir_entry->d_name);
		  	            else {
			                fprintf(stderr,"Couldn't delete %s. ", dir_entry->d_name);
			                perror("");
			            }
			        } else
			            if (verbose) printf(" Done\n");
			    }
		        closedir(dir);
		    }
		}
		if(chdir(current_dir) != 0) {
		    fprintf(stderr, "Error: Can't chdir to %s (aborting) - ", current_dir);
		    perror("");
		    exit(1);
		}
/*
		lstat(current_dir, &controlstat);
		if ( (cwd_stat.st_dev != controlstat.st_dev) || (cwd_stat.st_ino != controlstat.st_ino) ) {
                    fprintf(stderr, "Race found! (directory %s was exchanged or its your working directory)\n", current_dir);
                    exit(1);
                }
*/
		i_am_a_directory = 1;
	    }
        }
/* end of recursive function */


    if (S_ISREG(filestat.st_mode)) {

       /* open the file for writing in sync. mode */
        if ((fd = open(filename, O_RDWR | O_LARGEFILE | slow)) < 0) {
            /* here again this has a race problem ... hmmm */
            /* make it writable for us if possible */
            (void) chmod(filename, 0600);   /* ignore errors */
            if ((fd = open(filename, O_RDWR | O_LARGEFILE | slow)) < 0)
    	        return 1;
        }

        fstat(fd, &controlstat);
        if ((filestat.st_dev != controlstat.st_dev) || (filestat.st_ino != controlstat.st_ino) || (! S_ISREG(controlstat.st_mode))) {
            close(fd);
            return 3;
        }

        if (sdel_overwrite(mode, fd, 0, bufsize, filestat.st_size > 0 ? filestat.st_size : 1, zero) == 0)
            return sdel_unlink(filename, 0, 1, slow);
    } /* end IS_REG() */
    else {
        if (S_ISDIR(filestat.st_mode)) {
            if (i_am_a_directory == 0) {
                fprintf(stderr,"Warning: %s is a directory. I will not remove it, because the -r option is missing!\n", filename);
                return 0;
            } else
                return sdel_unlink(filename, 1, 0, slow);
        } else
            if (! S_ISDIR(filestat.st_mode)) {
	        fprintf(stderr,"Warning: %s is not a regular file, rename/unlink only!", filename);
	        if (! verbose)
	            printf("\n");
	        return sdel_unlink(filename, 0, 0, slow);
            }
    }

    return 99; // not reached
}

void cleanup(int signo) {
    fprintf(stderr,"Terminated by signal. Clean exit.\n");
    if (fd >= 0)
    	close(fd);
    FLUSH;
    exit(1);
}

int main (int argc, char *argv[]) {
    int errors = 0;
    int dot = 0;
    int result;
    int secure = 2; /* Standard is now SECURE mode (38 overwrites) [since v2.0] */
    int loop;
    struct rlimit rlim;

    prg = argv[0];
    if (argc < 2 || strncmp(argv[1], "-h", 2) == 0|| strncmp(argv[1], "--h", 3) == 0)
        help();

    while (1) {
        result = getopt(argc, argv, "DdFfLlRrSsVvZz");
        if (result < 0) break;
        switch (result) {
            case 'd' :
            case 'D' : dot = 1;
                       break;
            case 'F' :
            case 'f' : slow = 0;
            	       break;
	    case 'L' :
	    case 'l' : if (secure) secure--;
	    	       break;
	    case 'R' :
	    case 'r' : recursive++;
	    	       break;
	    case 'S' :
            case 's' : secure++;
            	       break;
	    case 'V' :
            case 'v' : verbose++;
            	       break;
            case 'Z' :
            case 'z' : zero++;
                       break;
            default :  help();       
        }
    }
    loop = optind;
    if (loop == argc)
        help();

    signal(SIGINT, cleanup);
    signal(SIGTERM, cleanup);
    signal(SIGHUP, cleanup);

    sdel_init(slow);

    if (verbose) {
        char type[15] = "random";
        if (zero) strcpy(type, "zero");
        switch (secure) {
           case 0 : printf("Wipe mode is insecure (one pass [%s])\n",type);
		    break;
	   case 1 : printf("Wipe mode is insecure (two passes [0xff/%s])\n",type);
	   	    break;
	   default: printf("Wipe mode is secure (38 special passes)\n");
	}
    }

#ifdef RLIM_INFINITY
 #ifdef RLIMIT_FSIZE
    rlim.rlim_cur = RLIM_INFINITY;
    rlim.rlim_max = RLIM_INFINITY;
    if (setrlimit(RLIMIT_FSIZE, &rlim) != 0)
    	fprintf(stderr, "Warning: Could not reset ulimit for filesize.\n");
 #else
    fprintf(stderr, "Warning: Not compiled with support for resetting ulimit filesize.\n");
 #endif
#endif

    while (loop < argc) {
        char rmfile[strlen(argv[loop]) + 1];
        strcpy(rmfile, argv[loop]);
        loop++;
        if (strcmp("/", rmfile) == 0) {
            fprintf(stderr,"Warning: Do you really want to remove the ROOT directory??\n");
            fprintf(stderr,"I'm giving you 5 seconds to abort ... press Control-C\n");
	    sleep(6);
            fprintf(stderr,"Doing my evil work now, don't whimp later, you had been informed!\n");
        }
        if (dot)
            if ((strcmp(".", rmfile) == 0) || (strcmp("..", rmfile) == 0))
                continue;
        if (verbose)
            printf("Wiping %s ", rmfile);
        result = (int) smash_it(rmfile, secure);
        switch (result) {
            case 0 : if (verbose) printf(" Done\n");
                     break;
            case 1 : fprintf(stderr, "Error: File %s - ", rmfile);
                     perror("");
                     break;
            case -1: break;
            case 3 : fprintf(stderr, "File %s was raced, hence I won't wipe it!\n", rmfile);
                     break;
            default: fprintf(stderr, "Unknown error\n");
        }
        if (result)
            errors++;
    }

    sdel_finnish();

    if (errors)
        return 1;
    else
        return 0;
}
